File: crates/maximus-checks/src/config_duplicates.rs
use maximus_core::{make_finding, FileKind, FindingInput, ProjectSnapshot, Severity};

use crate::check_outcome::CheckOutcome;
use crate::registry::{has_object_key, package_file_for_directory, read_package_json};

pub fn run_config_duplicate_check(project: &ProjectSnapshot) -> std::io::Result<CheckOutcome> {
    let mut findings = Vec::new();

    for directory in &project.directories {
        let package_file = package_file_for_directory(directory).cloned();
        let package_json = package_file
            .as_ref()
            .and_then(|file| read_package_json(&file.path));

        for (label, file_kind, package_field) in [
            ("ESLint", FileKind::Eslint, "eslintConfig"),
            ("Prettier", FileKind::Prettier, "prettier"),
            ("Jest", FileKind::Jest, "jest"),
        ] {
            let family_files = directory.files_by_kind.get(&file_kind);
            let family_file_count = family_files.map(|files| files.len()).unwrap_or(0);
            let has_package_field = package_json
                .as_ref()
                .map(|value| has_object_key(value, package_field))
                .unwrap_or(false);
            let total_sources = family_file_count + usize::from(has_package_field);

            if total_sources <= 1 {
                continue;
            }

            let severity = if label == "ESLint" {
                Severity::Error
            } else {
                Severity::Warn
            };
            let file = package_file
                .as_ref()
                .map(|value| value.path.clone())
                .or_else(|| family_files.and_then(|files| files.first().map(|file| file.path.clone())));

            findings.push(make_finding(FindingInput {
                id: format!("duplicate-config:{label}:{}", directory.dir.to_string_lossy()),
                title: format!("{label} config is declared in multiple places"),
                category: Some("duplicates".to_string()),
                detail: Some(format!(
                    "Found {total_sources} {label} config sources in {}.",
                    js_detail_directory(&directory.relative_dir)
                )),
                file,
                fix_ids: Vec::new(),
                fixable: false,
                hint: Some(format!(
                    "Keep a single {label} entry point per directory to avoid drift."
                )),
                severity: Some(severity),
            }));
        }

        let eslint_files = directory
            .files_by_kind
            .get(&FileKind::Eslint)
            .cloned()
            .unwrap_or_default();
        let has_legacy_eslint = eslint_files
            .iter()
            .any(|file| file.name.starts_with(".eslintrc"));
        let has_flat_eslint = eslint_files
            .iter()
            .any(|file| file.name.starts_with("eslint.config."));

        if has_legacy_eslint && has_flat_eslint {
            findings.push(make_finding(FindingInput {
                id: format!("eslint-mixed-modes:{}", directory.dir.to_string_lossy()),
                title: "Legacy and flat ESLint configs coexist".to_string(),
                category: Some("duplicates".to_string()),
                detail: Some(
                    "ESLint may resolve different config systems depending on invocation and toolchain."
                        .to_string(),
                ),
                file: eslint_files.first().map(|file| file.path.clone()),
                fix_ids: Vec::new(),
                fixable: false,
                hint: Some(
                    "Pick either flat config (eslint.config.*) or legacy .eslintrc.* in the same directory."
                        .to_string(),
                ),
                severity: Some(Severity::Error),
            }));
        }
    }

    Ok(CheckOutcome {
        findings,
        fixes: Vec::new(),
        planned_fixes: Vec::new(),
    })
}

fn js_detail_directory(relative_dir: &str) -> String {
    if cfg!(windows) {
        relative_dir.replace('/', "\\")
    } else {
        relative_dir.to_string()
    }
}

File: src/checks/config-duplicates.js
import { readTextIfExists } from "../lib/fs.js";
import { parseJsonc } from "../lib/jsonc.js";
import { makeFinding } from "../core/findings.js";
import { getDirectories } from "../core/discover.js";

const CONFIG_FAMILIES = [
  { label: "ESLint", fileKind: "eslint", packageField: "eslintConfig" },
  { label: "Prettier", fileKind: "prettier", packageField: "prettier" },
  { label: "Jest", fileKind: "jest", packageField: "jest" },
];

export async function runConfigDuplicateCheck(project) {
  const findings = [];

  for (const directory of getDirectories(project)) {
    const packageFile = directory.files.find((file) => file.kind === "package");
    const packageJson = packageFile ? await readPackageJson(packageFile.path) : null;

    for (const family of CONFIG_FAMILIES) {
      const familyFiles = directory.filesByKind.get(family.fileKind) ?? [];
      const hasPackageField = packageJson && Object.hasOwn(packageJson, family.packageField);
      const totalSources = familyFiles.length + (hasPackageField ? 1 : 0);

      if (totalSources <= 1) {
        continue;
      }

      findings.push(
        makeFinding({
          id: `duplicate-config:${family.label}:${directory.dir}`,
          category: "duplicates",
          severity: family.label === "ESLint" ? "error" : "warn",
          title: `${family.label} config is declared in multiple places`,
          file: packageFile?.path ?? familyFiles[0]?.path ?? null,
          detail: `Found ${totalSources} ${family.label} config sources in ${directory.relativeDir}.`,
          hint: `Keep a single ${family.label} entry point per directory to avoid drift.`,
        }),
      );
    }

    const eslintFiles = directory.filesByKind.get("eslint") ?? [];
    const hasLegacyEslint = eslintFiles.some((file) => file.name.startsWith(".eslintrc"));
    const hasFlatEslint = eslintFiles.some((file) => file.name.startsWith("eslint.config."));

    if (hasLegacyEslint && hasFlatEslint) {
      findings.push(
        makeFinding({
          id: `eslint-mixed-modes:${directory.dir}`,
          category: "duplicates",
          severity: "error",
          title: "Legacy and flat ESLint configs coexist",
          file: eslintFiles[0]?.path ?? null,
          detail:
            "This directory contains both legacy .eslintrc.* files and flat eslint.config.* files, so ESLint can resolve different rule sets depending on the entry point.",
          hint:
            "Migrate to eslint.config.* as the single source of truth, then remove the legacy .eslintrc.* files after the new config fully replaces them.",
        }),
      );
    }
  }

  return { findings, fixes: [] };
}

async function readPackageJson(filePath) {
  try {
    const text = await readTextIfExists(filePath);
    if (!text) {
      return null;
    }

    return parseJsonc(text, filePath);
  } catch {
    return null;
  }
}
